nestkernel/stimulating_device.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ *  stimulating_device.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+
+// Includes from nestkernel:
+#include "stimulating_device.h"
+#include "kernel_manager.h"
+
+
+nest::StimulatingDevice::StimulatingDevice()
+  : DeviceNode()
+  , Device()
+  , first_syn_id_( invalid_synindex )
+  , backend_params_( new Dictionary )
+{
+}
+
+nest::StimulatingDevice::StimulatingDevice( StimulatingDevice const& sd )
+  : DeviceNode( sd )
+  , Device( sd )
+  , first_syn_id_( invalid_synindex ) // a new instance can have no connections
+  , backend_params_( sd.backend_params_ )
+{
+  P_ = Parameters_( sd.P_ );
+}
+
+bool
+nest::StimulatingDevice::is_active( const Time& T ) const
+{
+  long step = T.get_steps();
+  if ( get_type() == StimulatingDevice::Type::CURRENT_GENERATOR
+    || get_type() == StimulatingDevice::Type::DELAYED_RATE_CONNECTION_GENERATOR
+    || get_type() == StimulatingDevice::Type::DOUBLE_DATA_GENERATOR )
+  {
+    step = step + 2;
+  }
+  return get_t_min_() < step and step <= get_t_max_();
+}
+
+void
+nest::StimulatingDevice::enforce_single_syn_type( synindex syn_id )
+{
+  if ( first_syn_id_ == invalid_synindex )
+  {
+    first_syn_id_ = syn_id;
+  }
+  if ( syn_id != first_syn_id_ )
+  {
+    throw IllegalConnection(
+      "All outgoing connections from a device must use the same synapse "
+      "type." );
+  }
+}
+
+void
+nest::StimulatingDevice::calibrate()
+{
+  Device::calibrate();
+}
+
+void
+nest::StimulatingDevice::set_initialized_()
+{
+  kernel().io_manager.enroll_stimulator( P_.stimulus_source_, *this, backend_params_ );
+}
+
+const std::string&
+nest::StimulatingDevice::get_label() const
+{
+  return P_.label_;
+}
+
+
+nest::StimulatingDevice::Parameters_::Parameters_()
+  : label_()
+  , stimulus_source_( names::internal )
+{
+}
+
+nest::StimulatingDevice::Parameters_::Parameters_( const Parameters_& p ) = default;
```

Comment:
The copy constructor should be defaulted in the declaration in the header file, not here.

Reply:
Done